README
# jlu-drcom-gtk
JLU Drcom Gtk - Client to login to Jilin University network

## Manual work
Please change the mac address in drcom.c before compiling

## How to compile
gcc jdg-main.c drcom.c md5.c -o jlu-drcom-gtk `pkg-config --cflags --libs gtk+-3.0 glib-2.0`

## TODO
* Automatically read mac address
* Add logout functionality


NEWS
Initial release.
...


drcom.c
/*
 *  simple JLU drcom client
 *  dirty hack version
 *
 */

/*
 * -- doc --
 * 数据包类型表示
 * 0x01 challenge request
 * 0x02 challenge response
 * 0x03 login request
 * 0x04 login response
 * 0x07 keep_alive request
 *		keep_alive response
 *		logout request
 *		logout response
 *		change_pass request
 *		change_pass response
 *
 */

#include <stdio.h>
#include <stdint.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>

#include <glib.h>

#include "drcom.h"
#include "md5.h"

void
get_user_info (struct user_info_pkt *user_info,
               const char *username,
               const char *password)
{
    user_info->username = username;
    user_info->username_len = strlen (username);
    user_info->password = password;
    user_info->password_len = strlen (password);
    user_info->hostname = host;
    user_info->hostname_len = host_len;
    user_info->os_name = os;
    user_info->os_name_len = os_len;
    user_info->mac_addr = mac;
}

void
set_challenge_data (unsigned char *clg_data,
                    int clg_data_len,
                    int clg_try_count)
{
    /* set challenge */
    int random = rand () % 0xF0 + 0xF;
    int data_index = 0;

    memset (clg_data, 0x00, clg_data_len);

    /* 0x01 challenge request */
    clg_data[data_index++] = 0x01;
    /* clg_try_count first 0x02, then increment */
    clg_data[data_index++] = 0x02 + (unsigned char)clg_try_count;
    /* two byte of challenge_data */
    clg_data[data_index++] = (unsigned char)(random % 0xFFFF);
    clg_data[data_index++] = (unsigned char)((random % 0xFFFF) >> 8);
    /* end with 0x0a */
    clg_data[data_index++] = 0x0a;
}

bool
challenge (int sock,
           struct sockaddr_in serv_addr,
           unsigned char *clg_data,
           int clg_data_len,
           char *recv_data,
           int recv_len,
           GtkTextBuffer *text_buffer)
{
    int ret;
    int challenge_try = 0;

    do
    {
        if (challenge_try > CHALLENGE_TRY)
        {
            close (sock);
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-challenge]: try challenge, but failed, please check your network connection.\n",
                                              strlen ("[drcom-challenge]: try challenge, but failed, please check your network connection.\n"));
            fprintf (stdout,
                     "[drcom-challenge]: try challenge, but failed, please check your network connection.\n");

            return false;
        }

        set_challenge_data (clg_data, clg_data_len, challenge_try);
        challenge_try++;

        ret = sendto (sock,
                      clg_data, clg_data_len, 0,
                      (struct sockaddr *)&serv_addr, sizeof (serv_addr));
        if (ret != clg_data_len)
        {
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-challenge]: send challenge data failed\n",
                                              strlen ("[drcom-challenge]: send challenge data failed\n"));
            fprintf (stdout, "[drcom-challenge]: send challenge data failed\n");

            continue;
        }

        ret = recvfrom (sock, recv_data, recv_len, 0, NULL, NULL);
        if (ret < 0)
        {
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-challenge]: recieve data from server failed.\n",
                                              strlen ("[drcom-challenge]: recieve data from server failed.\n"));
            fprintf (stdout,
                     "[drcom-challenge]: recieve data from server failed.\n");

            continue;
        }

        if (*recv_data != 0x02)
        {
            if (*recv_data == 0x07)
            {
                close (sock);

                gtk_text_buffer_insert_at_cursor (text_buffer,
                                                  "[drcom-challenge]: wrong challenge data.\n",
                                                  strlen ("[drcom-challenge]: wrong challenge data.\n"));
                fprintf (stdout, "[drcom-challenge]: wrong challenge data.\n");

                return false;
            }

            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-challenge]: challenge failed!, try again.\n",
                                              strlen ("[drcom-challenge]: challenge failed!, try again.\n"));
            fprintf (stdout, "[drcom-challenge]: challenge failed!, try again.\n");
        }
    } while ((*recv_data != 0x02));

    gtk_text_buffer_insert_at_cursor (text_buffer,
                                      "[drcom-challenge]: challenge success!\n",
                                      strlen ("[drcom-challenge]: challenge success!\n"));
    fprintf (stdout, "[drcom-challenge]: challenge success!\n");

    return true;
}

void
set_login_data (struct user_info_pkt *user_info,
                unsigned char *login_data,
                int login_data_len,
                unsigned char *salt,
                int salt_len)
{
    /* login data */
    int i, j;
    unsigned char md5_str[16];
    unsigned char md5_str_tmp[100];
    int md5_str_len;
    int data_index = 0;

    memset (login_data, 0x00, login_data_len);

    /* magic 3 byte, username_len 1 byte */
    login_data[data_index++] = 0x03;
    login_data[data_index++] = 0x01;
    login_data[data_index++] = 0x00;
    login_data[data_index++] = (unsigned char)(user_info->username_len + 20);

    /* md5 0x03 0x01 salt password */
    md5_str_len = 2 + salt_len + user_info->password_len;
    memset (md5_str_tmp, 0x00, md5_str_len);
    md5_str_tmp[0] = 0x03;
    md5_str_tmp[1] = 0x01;
    memcpy (md5_str_tmp +2, salt, salt_len);
    memcpy (md5_str_tmp + 2 + salt_len,
            user_info->password,
            user_info->password_len);
    MD5 (md5_str_tmp, md5_str_len, md5_str);
    memcpy (login_data + data_index, md5_str, 16);
    data_index += 16;

    /* user name 36 */
    memcpy (login_data + data_index,
            user_info->username,
            user_info->username_len);
    data_index += user_info->username_len > 36 ? user_info->username_len : 36;

    /* 0x00 0x00 */
    data_index += 2;

    /* (data[4:10].encode('hex'),16)^mac */
    uint64_t sum = 0;
    for (i = 0; i < 6; i++)
    {
        sum = (int)md5_str[i] + sum * 256;
    }
    sum ^= user_info->mac_addr;
    for (i = 6; i > 0; i--)
    {
        login_data[data_index + i - 1] = (unsigned char)(sum % 256);
        sum /= 256;
    }
    data_index += 6;

    /* md5 0x01 pwd salt 0x00 0x00 0x00 0x00 */
    md5_str_len = 1 + user_info->password_len + salt_len + 4;
    memset (md5_str_tmp, 0x00, md5_str_len);
    md5_str_tmp[0] = 0x01;
    memcpy (md5_str_tmp + 1, user_info->password, user_info->password_len);
    memcpy (md5_str_tmp + 1 + user_info->password_len, salt, salt_len);
    MD5 (md5_str_tmp, md5_str_len, md5_str);
    memcpy (login_data + data_index, md5_str, 16);
    data_index += 16;

    /* 0x01 0x31 0x8c 0x21 0x28 0x00*12 */
    login_data[data_index++] = 0x01;
    login_data[data_index++] = 0x31;
    login_data[data_index++] = 0x8c;
    login_data[data_index++] = 0x21;
    login_data[data_index++] = 0x28;
    data_index += 12;

    /* md5 login_data[0-data_index] 0x14 0x00 0x07 0x0b 8 bytes */
    md5_str_len = data_index + 4;
    memset (md5_str_tmp, 0x00, md5_str_len);
    memcpy (md5_str_tmp, login_data, data_index);
    md5_str_tmp[data_index+0] = 0x14;
    md5_str_tmp[data_index+1] = 0x00;
    md5_str_tmp[data_index+2] = 0x07;
    md5_str_tmp[data_index+3] = 0x0b;
    MD5 (md5_str_tmp, md5_str_len, md5_str);
    memcpy (login_data + data_index, md5_str, 8);
    data_index += 8;

    /* 0x01 0x00*4 */
    login_data[data_index++] = 0x01;
    data_index += 4;

    /* hostname */
    i = user_info->hostname_len > 71 ? 71 : user_info->hostname_len;
    memcpy (login_data + data_index, user_info->hostname, i);
    data_index += 71;

    /* 0x01 */
    login_data[data_index++] = 0x01;

    /* osname */
    i = user_info->os_name_len > 128 ? 128 : user_info->os_name_len;
    memcpy (login_data + data_index, user_info->os_name, i);
    data_index += 128;

    /* 0x6d 0x00 0x00 len(pass) */
    login_data[data_index++] = 0x6d;
    login_data[data_index++] = 0x00;
    login_data[data_index++] = 0x00;
    login_data[data_index++] = (unsigned char)(user_info->password_len);

    /* ror (md5 0x03 0x01 salt pass) pass */
    md5_str_len = 2 + salt_len + user_info->password_len;
    memset (md5_str_tmp, 0x00, md5_str_len);
    md5_str_tmp[0] = 0x03;
    md5_str_tmp[1] = 0x01;
    memcpy (md5_str_tmp +2, salt, salt_len);
    memcpy (md5_str_tmp + 2 + salt_len,
            user_info->password,
            user_info->password_len);
    MD5 (md5_str_tmp, md5_str_len, md5_str);
    int ror_check = 0;
    for (i = 0; i < user_info->password_len; i++)
    {
        ror_check = (int)md5_str[i] ^ (int)(user_info->password[i]);
        login_data[data_index++] = (unsigned char)(((ror_check << 3) & 0xFF) + (ror_check >> 5));
    }

    /* 0x02 0x0c */
    login_data[data_index++] = 0x02;
    login_data[data_index++] = 0x0c;

    /* checksum point */
    int check_point = data_index;
    login_data[data_index++] = 0x01;
    login_data[data_index++] = 0x26;
    login_data[data_index++] = 0x07;
    login_data[data_index++] = 0x11;

    /* 0x00 0x00 mac */
    login_data[data_index++] = 0x00;
    login_data[data_index++] = 0x00;
    uint64_t mac = user_info->mac_addr;
    for (i = 0; i < 6; i++)
    {
        login_data[data_index + i - 1] = (unsigned char)(mac % 256);
        mac /= 256;
    }
    data_index += 6;

    /* 0x00 0x00 0x00 0x00 the last two byte I dont't know*/
    login_data[data_index++] = 0x00;
    login_data[data_index++] = 0x00;
    login_data[data_index++] = 0x00;
    login_data[data_index++] = 0x00;

    /* checksum */
    sum = 1234;
    uint64_t check = 0;
    for (i = 0; i < data_index; i += 4)
    {
        check = 0;
        for (j = 0; j < 4; j++)
        {
            check = check * 256 + (int)login_data[i+j];
        }
        sum ^= check;
    }
    sum = (1968 * sum) & 0xFFFFFFFF;
    for (j = 0; j < 4; j++)
    {
        login_data[check_point+j] = (unsigned char)(sum >> (j*8) & 0x000000FF);
    }
}

bool
login (int sock,
       struct sockaddr_in serv_addr,
       unsigned char *login_data,
       int login_data_len,
       char *recv_data,
       int recv_len,
       GtkTextBuffer *text_buffer)
{
    /* login */
    int ret = 0;
    int login_try = 0;

    do
    {
        if (login_try > LOGIN_TRY)
        {
            close (sock);
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-login]: try login, but failed, something wrong\n",
                                              strlen ("[drcom-login]: try login, but failed, something wrong\n"));
            fprintf (stdout,
                     "[drcom-login]: try login, but failed, something wrong\n");

            return false;
        }

        login_try++;

        ret = sendto (sock,
                      login_data, login_data_len, 0,
                      (struct sockaddr *)&serv_addr, sizeof (serv_addr));
        if (ret != login_data_len)
        {
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-login]: send login data failed.\n",
                                              strlen ("[drcom-login]: send login data failed.\n"));
            fprintf (stdout, "[drcom-login]: send login data failed.\n");

            continue;
        }

        ret = recvfrom (sock, recv_data, recv_len, 0, NULL, NULL);
        if (ret < 0)
        {
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-login]: recieve data from server failed.\n",
                                              strlen ("[drcom-login]: recieve data from server failed.\n"));
            fprintf(stdout,
                    "[drcom-login]: recieve data from server failed.\n");

            continue;
        }

        if (*recv_data != 0x04)
        {
            if (*recv_data == 0x05)
            {
                close (sock);
                gtk_text_buffer_insert_at_cursor (text_buffer,
                                                  "[drcom-login]: wrong password or username!\n\n",
                                                  strlen ("[drcom-login]: wrong password or username!\n\n"));
                fprintf (stdout,
                         "[drcom-login]: wrong password or username!\n\n");
                return false;
            }
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom-login]: login failed!, try again\n",
                                              strlen ("[drcom-login]: login failed!, try again\n"));
            fprintf (stdout, "[drcom-login]: login failed!, try again\n");
        }
    } while ((*recv_data != 0x04));

    gtk_text_buffer_insert_at_cursor (text_buffer,
                                      "[drcom-login]: login success!\n",
                                      strlen ("[drcom-login]: login success!\n"));
    fprintf (stdout, "[drcom-login]: login success!\n");

    return true;
}

void
set_alive_data (unsigned char *alive_data,
                int alive_data_len,
                unsigned char *tail,
                int tail_len,
                int alive_count,
                int random)
{
    // 0: 84 | 1: 82 | 2: 82
    int i = 0;

    memset (alive_data, 0x00, alive_data_len);

    alive_data[i++] = 0x07;
    alive_data[i++] = (unsigned char)alive_count;
    alive_data[i++] = 0x28;
    alive_data[i++] = 0x00;
    alive_data[i++] = 0x0b;
    alive_data[i++] = (unsigned char)(alive_count * 2 + 1);
//  if (alive_count)
//  {
        alive_data[i++] = 0xdc;
        alive_data[i++] = 0x02;
//  }
//  else
//  {
//      alive_data[i++] = 0x0f;
//      alive_data[i++] = 0x27;
//  }

    random += rand() % 10;
    for (i = 9 ; i > 7; i--)
    {
        alive_data[i] = random % 256;
        random /= 256;
    }
    memcpy (alive_data + 16, tail, tail_len);
    i = 25;
//  if (alive_count && alive_count % 3 == 0)
//  {
//      /* crc */
//      memset (alive_data, 0xFF, 16);
//  }
}

void
set_logout_data (unsigned char *logout_data,
                 int logout_data_len)
{
    memset (logout_data, 0x00, logout_data_len);
    // TODO
}

bool
logout (int sock,
        struct sockaddr_in serv_addr,
        unsigned char *logout_data,
        int logout_data_len,
        char *recv_data,
        int recv_len,
        GtkTextBuffer *text_buffer)
{
    set_logout_data (logout_data, logout_data_len);
    // TODO

    close (sock);
    return false;
}

void
logout_signal (int signum)
{
    logout_flag = 1;
}

gboolean
keep_alive (gpointer user_data)
{
    // keep alive alive data length 42 or 40
    unsigned char tail[4];
    int tail_len = 4;
    int alive_data_len = 0;
    int ret;
    int sock;
    struct sockaddr_in serv_addr;
    unsigned char send_data[SEND_DATA_SIZE];
    char recv_data[RECV_DATA_SIZE];

    sock = *((int *)user_data);

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(SERVER_ADDR);
    serv_addr.sin_port = htons(SERVER_PORT);

    memset (tail, 0x00, tail_len);

alive:
    if (alive_fail_count > ALIVE_TRY)
    {
        close (sock);
        fprintf (stdout,
                 "[drcom-keep-alive]: couldn't connect to network, check please.\n");

        return G_SOURCE_REMOVE;
    }

    alive_data_len = alive_count > 0 ? 40 : 42;
    set_alive_data (send_data,
                    alive_data_len, tail, tail_len, alive_count, random_num);

    ret = sendto (sock,
                  send_data, alive_data_len, 0,
                  (struct sockaddr *)&serv_addr, sizeof (serv_addr));
    if (ret != alive_data_len)
    {
        alive_fail_count++;

        fprintf (stdout,
                 "[drcom-keep-alive]: send keep-alive data failed.\n");

        goto alive;
    }
    else
    {
        alive_fail_count = 0;
    }

    memset (recv_data, 0x00, RECV_DATA_SIZE);

    ret = recvfrom (sock, recv_data, RECV_DATA_SIZE, 0, NULL, NULL);
    if (ret < 0 || *recv_data != 0x07)
    {
        alive_fail_count++;

        fprintf (stdout,
                 "[drcom-keep-alive]: recieve keep-alive response data from server failed.\n");

        goto alive;
    }
    else
    {
        alive_fail_count = 0;
    }

    if (alive_count > 1)
    {
        memcpy (tail, recv_data+16, tail_len);
    }

    fprintf (stdout, "[drcom-keep-alive]: keep alive.\n");
    alive_count = (alive_count + 1) % 3;

    return G_SOURCE_CONTINUE;
}


drcom.h
/*
 *  JLU Drcom Gtk - Client to login to Jilin University network
 *  Copyright (C) 2016  Jonathan Kang
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef DRCOM_H_
#define DRCOM_H_

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdbool.h>
#include <netinet/in.h>

#include <gtk/gtk.h>

// 必须修改，帐号密码和 mac 地址是绑定的
static uint64_t mac = 0x000000000000; // echo 0x`ifconfig eth | egrep -io "([0-9a-f]{2}:){5}[0-9a-f]{2}" | tr -d ":"`
static int random_num;
static int alive_count = 0;
static int alive_fail_count = 0;

// 不一定要修改
static char host[] = "drcom";
static char os[] = "drcom";
static int host_len = sizeof(host) - 1;
static int os_len = sizeof(os) - 1;

/* signal process flag */
static int logout_flag = 0;

// TODO 增加从文件读取参数

//SERVER_DOMAIN login.jlu.edu.cn
#define SERVER_ADDR "10.100.61.3"
#define SERVER_PORT 61440

#define RECV_DATA_SIZE 1000
#define SEND_DATA_SIZE 1000
#define CHALLENGE_TRY 10
#define LOGIN_TRY 5
#define ALIVE_TRY 5

/* infomation */
struct user_info_pkt {
	const char *username;
	const char *password;
	char *hostname;
	char *os_name;
	uint64_t mac_addr;
	int username_len;
	int password_len;
	int hostname_len;
	int os_name_len;
};

void get_user_info (struct user_info_pkt *user_info,
                    const char *username,
                    const char *password);
void set_challenge_data (unsigned char *clg_data,
                         int clg_data_len,
                         int clg_try_count);
bool challenge (int sock,
                struct sockaddr_in serv_addr,
                unsigned char *clg_data,
                int clg_data_len,
                char *recv_data,
                int recv_len,
                GtkTextBuffer *text_buffer);
void set_login_data (struct user_info_pkt *user_info,
                     unsigned char *login_data,
                     int login_data_len,
                     unsigned char *salt,
                     int salt_len);
bool login (int sock,
            struct sockaddr_in serv_addr,
            unsigned char *login_data,
            int login_data_len,
            char *recv_data,
            int recv_len,
            GtkTextBuffer *text_buffer);
void set_alive_data (unsigned char *alive_data,
                     int alive_data_len,
                     unsigned char *tail,
                     int tail_len,
                     int alive_count,
                     int random);
void set_logout_data (unsigned char *logout_data,
                      int logout_data_len);
bool logout (int sock,
             struct sockaddr_in serv_addr,
             unsigned char *logout_data,
             int logout_data_len,
             char *recv_data,
             int recv_len,
             GtkTextBuffer *text_buffer);
void logout_signal (int signum);
gboolean keep_alive (gpointer user_data);
void on_login (const char *username,
               const char *password,
               GtkTextBuffer *text_buffer);

#endif /* DRCOM_H_ */


jdg-main.c
/*
 *  JLU Drcom Gtk - Client to login to Jilin University network
 *  Copyright (C) 2016  Jonathan Kang
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <gtk/gtk.h>
#include <glib-object.h>

#include <string.h>
#include <arpa/inet.h>

#include "drcom.h"

static int sock;
static guint source_id;

static void
on_destroy (GtkWidget *widget,
            gpointer user_data)
{
    GPtrArray *array = (GPtrArray *) user_data;

    g_ptr_array_free (array, FALSE);

    if (source_id)
    {
        g_source_remove (source_id);
        source_id = 0;
    }

    gtk_main_quit ();
}

static void
on_check_button_toggled (GtkWidget *check_button,
                         gpointer user_data)
{
    GKeyFile *key_file;

    key_file = g_key_file_new ();
    g_key_file_load_from_file (key_file, "jdg.ini",
                               G_KEY_FILE_NONE, NULL);

    if (gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (check_button)))
    {
        g_key_file_set_boolean (key_file,
                                "config", "remember_password", TRUE);
    }
    else
    {
        g_key_file_set_boolean (key_file,
                                "config", "remember_password", FALSE);

        if (g_key_file_has_key (key_file, "config", "username", NULL))
        {
            g_key_file_remove_key (key_file, "config", "username", NULL);
            g_key_file_remove_key (key_file, "config", "password", NULL);
        }
    }

    g_key_file_save_to_file (key_file, "jdg.ini", NULL);

    g_key_file_free (key_file);
}

static void
on_login_button_clicked (GtkWidget *button,
                         gpointer user_data)
{
    GPtrArray *array;
    GtkEntryBuffer *username_buffer;
    GtkEntryBuffer *password_buffer;

    array = (GPtrArray *) user_data;
    username_buffer = g_ptr_array_index (array, 0);
    password_buffer = g_ptr_array_index (array, 1);

    if (gtk_entry_buffer_get_length (username_buffer) == 0
        || gtk_entry_buffer_get_length (password_buffer) == 0)
    {
        GtkWidget *dialog;
        GtkWidget *toplevel;

        toplevel = gtk_widget_get_toplevel (button);
        dialog = gtk_message_dialog_new (GTK_WINDOW (toplevel),
                                         GTK_DIALOG_MODAL,
                                         GTK_MESSAGE_WARNING,
                                         GTK_BUTTONS_CLOSE,
                                         "Please input username or password");

        gtk_dialog_run (GTK_DIALOG (dialog));
        gtk_widget_destroy (dialog);
    }
    else
    {
        const char *username;
        const char *password;
        GKeyFile *key_file;
        GtkWidget *check_button;
        GtkWidget *revealer;
        GtkWidget *stack;
        GtkWidget *text_view;
        GtkTextBuffer *text_buffer;

        /* get username and password from entry buffers */
        username = gtk_entry_buffer_get_text (username_buffer);
        password = gtk_entry_buffer_get_text (password_buffer);

        check_button = g_ptr_array_index (array, 2);
        revealer = g_ptr_array_index (array, 3);
        stack = g_ptr_array_index (array, 4);
        text_view = g_ptr_array_index (array, 5);

        text_buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (text_view));

        key_file = g_key_file_new ();
        g_key_file_load_from_file (key_file, "jdg.ini",
                                   G_KEY_FILE_NONE, NULL);

        if (gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (check_button)))
        {
            if (g_key_file_has_key (key_file, "config", "username", NULL))
            {
                gchar *str_un;

                str_un = g_key_file_get_string (key_file,
                                                "config", "username", NULL);

                if (g_strcmp0 (username, str_un) != 0)
                {
                    g_key_file_set_string (key_file,
                                           "config", "username", username);
                    g_key_file_set_string (key_file,
                                           "config", "password", password);
                }

                g_free (str_un);
            }
            else
            {
                g_key_file_set_string (key_file,
                                       "config", "username", username);
                g_key_file_set_string (key_file,
                                       "config", "password", password);
            }
        }
        else
        {
            if (g_key_file_has_key (key_file, "config", "username", NULL))
            {
                g_key_file_remove_key (key_file, "config", "username", NULL);
                g_key_file_remove_key (key_file, "config", "password", NULL);
            }
        }

        g_key_file_save_to_file (key_file, "jdg.ini", NULL);

        g_key_file_free (key_file);

        gtk_revealer_set_reveal_child (GTK_REVEALER (revealer), TRUE);
        gtk_stack_set_visible_child_name (GTK_STACK (stack), "info");

        /* login */
        int ret;
        unsigned char send_data[SEND_DATA_SIZE];
        char recv_data[RECV_DATA_SIZE];
        struct sockaddr_in serv_addr;
        struct user_info_pkt user_info;

        sock = socket (AF_INET, SOCK_DGRAM, 0);
        if (sock < 0)
        {
            gtk_text_buffer_insert_at_cursor (text_buffer,
                                              "[drcom]: create sock failed.\n",
                                              strlen ("[drcom]: create sock failed.\n"));
            fprintf (stdout, "[drcom]: create sock failed.\n");
            //exit (EXIT_FAILURE);
        }
        serv_addr.sin_family = AF_INET;
        serv_addr.sin_addr.s_addr = inet_addr(SERVER_ADDR);
        serv_addr.sin_port = htons(SERVER_PORT);

        // get user information
        get_user_info (&user_info, username, password);

        // challenge data length 20
        challenge (sock, serv_addr,
                   send_data, 20, recv_data, RECV_DATA_SIZE, text_buffer);

        // login data length 338, salt length 4
        set_login_data (&user_info,
                        send_data, 338, (unsigned char *)(recv_data + 4), 4);
        memset (recv_data, 0x00, RECV_DATA_SIZE);
        login (sock, serv_addr,
               send_data, 338, recv_data, RECV_DATA_SIZE, text_buffer);

        random_num = rand () % 0xFFFF;
        keep_alive (&sock);
        source_id = g_timeout_add_full (G_PRIORITY_DEFAULT, 15000,
                                        keep_alive, &sock, NULL);
    }
}

static void
on_logout_button_clicked (GtkWidget *button,
                          gpointer user_data)
{
    GPtrArray *array;
    GtkEntryBuffer *password_buffer;
    GtkWidget *check_button;
    GtkWidget *revealer;
    GtkWidget *stack;

    array = (GPtrArray *) user_data;
    password_buffer = g_ptr_array_index (array, 1);
    check_button = g_ptr_array_index (array, 2);
    revealer = g_ptr_array_index (array, 3);
    stack = g_ptr_array_index (array, 4);

    gtk_revealer_set_reveal_child (GTK_REVEALER (revealer), FALSE);
    gtk_stack_set_visible_child_name (GTK_STACK (stack), "login");

    if (!gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (check_button)))
    {
        const gchar *password;

        password = gtk_entry_buffer_get_text (password_buffer);
        gtk_entry_buffer_delete_text (password_buffer, 0, strlen (password));
    }

    /* stop keep_alive */
    if (source_id)
    {
        g_source_remove (source_id);
        source_id = 0;
    }
}

int
main (int argc, char *argv[])
{
    GdkPixbuf *icon;
    GtkWidget *window;
    GtkWidget *main_box;
    GtkWidget *grid;
    GtkWidget *username_label;
    GtkWidget *password_label;
    GtkWidget *username_entry;
    GtkWidget *password_entry;
    GtkWidget *check_button;
    GtkWidget *login_button;
    GtkWidget *logout_button;
    GtkWidget *revealer;
    GtkWidget *stack;
    GtkWidget *text_view;
    GtkEntryBuffer *username_buffer;
    GtkEntryBuffer *password_buffer;
    GPtrArray *array;
    GKeyFile *key_file;

    gtk_init (&argc, &argv);

    window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title (GTK_WINDOW (window), "jlu-drcom-gtk");
    gtk_window_set_default_size (GTK_WINDOW (window), 600, 200);
    gtk_window_set_resizable (GTK_WINDOW (window), FALSE);

    icon = gdk_pixbuf_new_from_file ("jdg.png", NULL);
    gtk_window_set_icon (GTK_WINDOW (window), icon);

    username_label = gtk_label_new ("Username");
    password_label = gtk_label_new ("Password");

    username_buffer = gtk_entry_buffer_new (NULL, -1);
    password_buffer = gtk_entry_buffer_new (NULL, -1);
    username_entry = gtk_entry_new_with_buffer (username_buffer);
    password_entry = gtk_entry_new_with_buffer (password_buffer);
    gtk_entry_set_input_purpose (GTK_ENTRY (password_entry),
                                 GTK_INPUT_PURPOSE_PASSWORD);
    gtk_entry_set_visibility (GTK_ENTRY (password_entry), FALSE);

    check_button = gtk_check_button_new_with_label ("Remember Password");
    login_button = gtk_button_new_with_label ("Login");

    revealer = gtk_revealer_new ();
    gtk_revealer_set_transition_duration (GTK_REVEALER (revealer), 500);

    /* Stack to switch between login page and info page */
    stack = gtk_stack_new ();

    /* GtkTextView to show output information */
    text_view = gtk_text_view_new ();
    gtk_text_view_set_editable (GTK_TEXT_VIEW (text_view), FALSE);
    gtk_widget_set_size_request (text_view, -1, 200);
    gtk_widget_set_vexpand (text_view, TRUE);
    gtk_widget_set_visible (text_view, TRUE);

    gtk_container_add (GTK_CONTAINER (revealer), text_view);

    array = g_ptr_array_new ();
    g_ptr_array_add (array, (gpointer) username_buffer);
    g_ptr_array_add (array, (gpointer) password_buffer);
    g_ptr_array_add (array, (gpointer) check_button);
    g_ptr_array_add (array, (gpointer) revealer);
    g_ptr_array_add (array, (gpointer) stack);
    g_ptr_array_add (array, (gpointer) text_view);

    key_file = g_key_file_new ();
    g_key_file_load_from_file (key_file, "jdg.ini",
                               G_KEY_FILE_NONE, NULL);
    if (g_key_file_get_boolean (key_file, "config", "remember_password", NULL))
    {
        gchar *username;
        gchar *password;

        gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (check_button), TRUE);

        username = g_key_file_get_string (key_file,
                                          "config", "username", NULL);
        password = g_key_file_get_string (key_file,
                                          "config", "password", NULL);

        if (username != NULL && password != NULL)
        {
            gtk_entry_buffer_set_text (username_buffer, username, -1);
            gtk_entry_buffer_set_text (password_buffer, password, -1);
        }

        g_free (username);
        g_free (password);
    }
    else
    {
        gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (check_button), FALSE);
    }
    g_key_file_free (key_file);

    /* Login grid */
    grid = gtk_grid_new ();
    gtk_grid_set_row_spacing (GTK_GRID (grid), 5);
    gtk_grid_set_column_spacing (GTK_GRID (grid), 8);

    gtk_grid_attach (GTK_GRID (grid), username_label, 0, 0, 1, 1);
    gtk_grid_attach (GTK_GRID (grid), password_label, 0, 1, 1, 1);
    gtk_grid_attach (GTK_GRID (grid), username_entry, 1, 0, 1, 1);
    gtk_grid_attach (GTK_GRID (grid), password_entry, 1, 1, 1, 1);
    gtk_grid_attach (GTK_GRID (grid), check_button, 0, 2, 1, 1);
    gtk_grid_attach (GTK_GRID (grid), login_button, 1, 2, 1, 1);

    logout_button = gtk_button_new_with_label ("logout");
    gtk_widget_set_halign (logout_button, GTK_ALIGN_CENTER);
    gtk_widget_set_valign (logout_button, GTK_ALIGN_CENTER);

    gtk_stack_add_named (GTK_STACK (stack), grid, "login");
    gtk_stack_add_named (GTK_STACK (stack), logout_button, "info");
    gtk_stack_set_visible_child_name (GTK_STACK (stack), "login");

    main_box = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
    gtk_widget_set_margin_start (main_box, 5);
    gtk_widget_set_margin_end (main_box, 5);
    gtk_box_pack_start (GTK_BOX (main_box), stack,
                        TRUE, TRUE, 5);
    gtk_box_pack_end (GTK_BOX (main_box), revealer,
                      TRUE, TRUE, 5);

    gtk_container_add (GTK_CONTAINER (window), main_box);
    gtk_widget_show_all (window);

    /* Signals */
    g_signal_connect (window, "destroy", G_CALLBACK (on_destroy), array);
    g_signal_connect (password_entry, "activate",
                      G_CALLBACK (on_login_button_clicked), array);
    g_signal_connect (check_button, "toggled",
                      G_CALLBACK (on_check_button_toggled), NULL);
    g_signal_connect (login_button, "clicked",
                      G_CALLBACK (on_login_button_clicked), array);
    g_signal_connect (logout_button, "clicked",
                      G_CALLBACK (on_logout_button_clicked), array);

    g_object_unref (icon);

    gtk_main ();

    return 0;
}
